### PR TITLE
Filter the benign torch DataParallel scalar-gather warning in CI

### LINKED ISSUE #7018: CI emits `torch` UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars
### What

CI emits `UserWarning`s from `torch` in the SFT tests, in the **multi-GPU** slow tests job only:

```
tests/test_sft_trainer.py: 16 warnings
  /__w/trl/trl/.venv/lib/python3.12/site-packages/torch/autograd/function.py:625: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
    return super().apply(*args, **kwargs)  # type: ignore[misc]
```

CI run: https://github.com/huggingface/trl/actions/runs/33618846043/job/100210907376

Affected tests: 16 warnings, all from `tests/test_sft_trainer.py` (pytest collapses them to the file, since the same warning is recorded in more than ten locations).

The single-GPU job emits none of them.

The tests themselves pass; for now this is warning noise only.

### When it started

The warnings appeared in the multi-GPU slow tests job on 2026-08-20. They are absent from every earlier run of that job and present in every run since:

| commit | run | warnings |
|---|---|---|
| `ce01ad43` | https://github.com/huggingface/trl/actions/runs/32359523680 | 0 |
| `4d37e9e2` | https://github.com/huggingface/trl/actions/runs/32394938434 | 16 |

The two runs in between (`3f738b90`, `29b75df4`) died before pytest printed a warnings summary, see #6836.

Nothing changed in the installed dependency versions across that boundary.

### Next steps

Investigate the root cause, then decide whether anything is actionable on the TRL side and/or should be reported upstream.

### PR DESCRIPTION
**Pull request by albertvillanova**
This PR silences the benign `torch` `nn.DataParallel` scalar-loss gather `UserWarning` emitted by the SFT tests in the multi-GPU CI job.

Fix #7018.

### Motivation

Since the multi-GPU slow tests job moved to a runner with two visible devices, `tests/test_sft_trainer.py` emits 16 warnings:

```
/__w/trl/trl/.venv/lib/python3.12/site-packages/torch/autograd/function.py:625: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  return super().apply(*args, **kwargs)  # type: ignore[misc]
```

`Trainer` wraps the model in `nn.DataParallel` whenever more than one accelerator is visible and no distributed launcher is used. `DataParallel` then gathers each replica's output, and since the `loss` field is a 0-dim scalar on every replica, `torch.nn.parallel._functions.Gather` unsqueezes them into a vector and warns about it. The gather happens entirely inside `DataParallel`, which also chooses the dimension, so there is nothing actionable on the TRL side. The tests pass.

Not reported upstream: `nn.DataParallel` is soft-deprecated in favour of `DistributedDataParallel` and the warning has been unchanged for years, identical at `v2.13.0` and on `pytorch/main`. It is also arguably intentional, since the gathered shape does differ from a plain gather along dimension 0. The two existing reports (huggingface/transformers#852 and huggingface/transformers#14128) were both closed with no code change.

### Solution

Suppress the warning through `filterwarnings` in `pyproject.toml`, documented like the other upstream warning filters. The entry is tied to #6836 rather than to a dependency version, since the warning disappears once the multi-GPU CI no longer exercises `nn.DataParallel`.

### Changes

- Add a `filterwarnings` entry for the `torch` `DataParallel` scalar-gather `UserWarning`, with a comment explaining the cause, what triggers it, and when the entry can be removed

---

> [!NOTE]
> **Low Risk**
> Only changes pytest warning filters for CI; no library or training runtime behavior.
> 
> **Overview**
> Adds a **pytest `filterwarnings` ignore** in `pyproject.toml` for PyTorch’s benign `UserWarning` when `nn.DataParallel` gathers per-replica **scalar losses** (“Was asked to gather along dimension 0, but all input tensors were scalars…”).
> 
> This noise showed up after the **multi-GPU slow-test runner** gained two visible devices: Hugging Face `Trainer` wraps the model in `DataParallel` in that setup, and the gather happens inside PyTorch—not something TRL can fix. The entry is documented like other upstream filters, with tracking **#7018** and a note to drop it when CI no longer exercises `DataParallel` (**#6836**).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit deb9e45b89a6bca85189c1c1141ca969ea323032. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by qgallouedec — APPROVED** [review]
thanks

**Comment by albertvillanova** [comment]
### Confirmed in CI

The slow tests only run on `main` and only for `trl/**.py` / `examples/**.py` changes, so this PR's own checks cannot exercise the warning: all their test jobs run on a single-GPU runner, and the two-GPU "Distributed smoke tests" job shells out to `accelerate launch` subprocesses, so nothing is wrapped in `nn.DataParallel` in-process there.

Verified with the slow tests trigger temporarily relaxed (that tweak is not part of this PR):

| | run | multi-GPU result |
|---|---|---|
| without the filter | [33618846043](https://github.com/huggingface/trl/actions/runs/33618846043/job/100210907376) | `34 passed, 44 skipped, 2528 deselected, 21 xfailed, 16 warnings` |
| with the filter | [33639890910](https://github.com/huggingface/trl/actions/runs/33639890910) | `34 passed, 44 skipped, 2528 deselected, 21 xfailed` |

Identical test outcome counts, and the warnings summary is gone entirely. The 21 xfails are themselves the check that the warning was filtered rather than merely not triggered: `xfail_data_parallel` is conditional on `backend_device_count(torch_device) > 1`, so two devices really were visible and `nn.DataParallel` really was in use.

`torch` moved 2.13.0 -> 2.14.0 between the two runs, which does not affect the comparison: the emitting code in `torch/nn/parallel/_functions.py` is unchanged at `v2.14.0-rc1` and on `pytorch/main`.
